# PCIP_FILTER

<p align="center">
  <img src="https://github.com/GIGA-AnimalGenomics-BLV/PCIP/blob/master/WORKFLOW/FILTER.svg?raw=true" />
</p>

``PCIP_filter(minimap2PAF = NULL, targetName = NULL)``

* **minimap2PAF** [tibble] PAF alignment file generated by minimap2 and loaded with ``readPairwiseAlignmentFile``.
* **targetName** [character] Name of the TARGET chromosome.

# PCIP_GETBREAKPOINTS

<p align="center">
  <img src="https://github.com/GIGA-AnimalGenomics-BLV/PCIP/blob/master/WORKFLOW/BREAKPOINT.svg?raw=true" />
</p>

``PCIP_getBreakPoints(PAF = NULL, lengthTarget = NULL, targetName = NULL, gapAlignment = NA, distanceToLTR = NA, returnFILTEREDout = FALSE)``

## MAIN
* **PAF** [tibble] PAF alignment file filtered with ``PCIP_filter``.
* **lengthTarget** [numeric] Length of the TARGET chromosome (bp).
* **targetName** [character] Name of the TARGET chromosome.

## OPTIONS
* **gapAlignment** [numeric or NA] Maximum mean gap tolerated between substrings. NA deactivates this option. For debugging purpose. 
* **distanceToLTR** [numeric or NA] Exclude reads with alignment gaps near the LTR edge > distanceToLTR. NA deactivates this option. For debugging purpose.
* **returnFILTEREDout** [boolean] Directly returns reads with alignment gaps near the LTR edge > distanceToLTR. For debugging purpose.

# PCIP_SUMMARISE

<p align="center">
  <img src="https://github.com/GIGA-AnimalGenomics-BLV/PCIP/blob/master/WORKFLOW/SUMMARISE.svg?raw=true" />
</p>

``PCIP_summarise(PCIPbreakpoints = NULL, align = NULL, targetName = NULL, mergeISdistance = 200, mergeShearSiteDistance = 0, minDistanceLTR = 300, distanceFLAG = 1000, distance_report = 1000)``

## MAIN
* **PCIPbreakpoints** [tibble] Break points results generated with ``PCIP_getBreakPoints`.
* **align** [tibble] PAF alignment file generated by minimap2 and loaded with ``readPairwiseAlignmentFile``.
* **targetName** [character] Name of the TARGET chromosome.
* **mergeISdistance** [numeric] Window (bp, up/down) to regroup close viral-host junctions into an integration site.
* **mergeShearSiteDistance** [numeric] Window (bp, up/down) to collapse reads displaying the same shear site. 
* **distanceFLAG** [numeric] Flag integration sites located too close (within distanceFLAG, up/down).
* **distance_report** [numeric] When reporting the integration site - read ID table, associate to each IS any reads within a +/- distance_report bp window.

# FLAGS

<p align="center">
  <img src="https://github.com/GIGA-AnimalGenomics-BLV/PCIP/blob/master/WORKFLOW/FLAGS.svg?raw=true" />
</p>

# OUTPUTS

<p align="center">
  <img src="https://github.com/GIGA-AnimalGenomics-BLV/PCIP/blob/master/WORKFLOW/OUTPUT.svg?raw=true" />
</p>
